### PR TITLE
Add TensorRT-LLM support for inference backend

### PR DESCRIPTION
**Pull request by chungen04**
## Summary

Integrates TensorRT-LLM (v1.3.0rc18) as an inference backend for hidden-state capture, alongside the existing vLLM and SGLang backends. Validated end-to-end against SGLang on Qwen3-8B (EAGLE-3 / DFlash).

## Approach

TRT-LLM already ships the EAGLE3 capture machinery natively (`SaveHiddenStatesDecodingConfig` / `SaveHiddenStatesResourceManager`), so this integration is thin. It builds on that mode and adds a small patch that redirects captured aux + final hidden states to Mooncake instead of writing .pt files to disk.

## Changes

* `patches/trtllm/v1.3.0rc18/trtllm.patch`: env-gated
  (`TORCHSPEC_TRTLLM_MOONCAKE`) Mooncake redirect in `save_hidden_state.py`. 
  * Adds batched-prefill capture (upstream only stored a single request per forward) by de-interleaving the packed capture buffer per request, walking `context_requests` in native order to match how the executor flattens tokens.
  * Keys the store on the client request id + DP rank.
  * Forwards Mooncake/TorchSpec env to MPI workers.
  * relaxes the `max_batch_size=1` cap for hidden state capture.
  * Disables KV block reuse, overlap scheduler, and chunked prefill (all required for correct, full-prefill capture).
* `torchspec/inference/engine/trtllm_engine.py`: 
  * `TrtllmEngine`: Ray actor wrapping the PyTorch-backend `LLM`
  * Maps `RequestOutput.request_id` to Mooncake key using the same sanitizer as the patch.
* `factory.py` / config: 
  * Includes `"trtllm"` dispatch, `TrtllmConfig`, `trtllm_` flatten prefix.
  * Example config in `configs/trtllm_qwen3_8b.yaml`.
* `docker/trtllm/v1.3.0rc18/Dockerfile` + `justfile` branch + `[trtllm]` extra  
  * patches the installed package at build time with an assertion guard, and includes the CUDA-13 Mooncake wheel fix.
* `configs/trtllm_qwen3_8b.yaml` + README entry + `qwen3-8b-single-node` example wired to the shared launcher.

## Performance

Setting: Qwen3-8B, single node (4xB300, 2 inference (TP=2) + 2 training FSDP), open-perfectblend dataset (not re-generated), global batch 8, 10k steps, 512 samples split for eval. Each backend evaluates on its own captured eval cache.

Quality is verified, identical (eval accuracy, matched at every checkpoint):

| step | SGLang | TRT-LLM |
|---|---|---|
| 2000 | 0.496 | 0.495 |
| 6000 | 0.620 | 0.619 |
| 10000 | 0.650 | 0.650 |

For E2E Speed, TRT-LLM is ~3.6× faster (training-loop wall-clock):

| backend | wall-clock |
|---|---|
| SGLang | 18886 s (5h15m) |
| TRT-LLM | 5297 s (1h28m) |

SGLang is inference-bound. TRT-LLM speeds up inference and becomes training-bound under this setting.

There is also a DFlash example added, mirroring SGLang's example (For SGLang's training, see #126 for an encountered hang). The training results shows quality remains equivalent while wall clock time also mirrors, as with both engine the pipeline was training-bound.

E2E speed for DFlash:
| | SGLang DFlash | TRT-LLM DFlash | 
|---|---|---|
| wall-clock | **4173.5 s** (69.6 min) | **4320.4 s** (72.0 min)  |
| step rate | 2.91 step/s | 2.83 step/s|
| avg training throughput | 19.2 entries/s | 18.5 entries/s  |
| avg inference throughput | 19.3 entries/s | 18.6 entries/s  |

Training results for DFlash:
| step | SGL loss | TRT loss | SGL acc | TRT acc | SGL sim_acc_len | TRT sim_acc_len |
|---|---|---|---|---|---|---|
| 1000 | 4.9995 | 4.9752 | 0.1381 | 0.1396 | 0.60 | 0.62 |
| 2000 | 4.5755 | 4.5726 | 0.1638 | 0.1630 | 0.75 | 0.75 |
| 3000 | 4.3092 | 4.2956 | 0.1783 | 0.1811 | 0.84 | 0.85 |
| 4000 | 4.1083 | 4.0983 | 0.1951 | 0.1967 | 0.93 | 0.94 |
| 5000 | 3.9469 | 3.9286 | 0.2091 | 0.2108 | 1.01 | 1.02 |
| 6000 | 3.8038 | 3.7964 | 0.2187 | 0.2204 | 1.07 | 1.08 |
| 7000 | 3.6954 | 3.6888 | 0.2305 | 0.2322 | 1.13 | 1.13 |
| 8000 | 3.6041 | 3.5976 | 0.2391 | 0.2405 | 1.17 | 1.18 |
| 9000 | 3.5430 | 3.5406 | 0.2453 | 0.2462 | 1.21 | 1.21 |
| 10000 | 3.5059 | 3.5020| 0.2491 | 0.2505 | 1.23 | 1.23 |

## Test

```bash
just BACKEND=trtllm build
deploy/run_compare.sh trtllm \
  dataset.train_data_path=<train.jsonl> \
  dataset.eval_data_path=<eval.jsonl> dataset.eval_interval=1000 \
  training.micro_batch_size=4 training.num_train_steps=10000
```

## Note
Multimodal input is not yet supported. I plan to split the multimodal input support and the base TRTLLM support into two PRs.

### CODE REVIEW

File: examples/qwen3-8b-single-node/run.sh [hunk 1]
```diff
@@ -56,14 +65,13 @@ echo "Local IP: $LOCAL_IP"
 echo "Extra args: $*"
 echo "=============================================="
 
-# TODO: unify tp_size config across sglang/vllm backends
```
Comment (chungen04) [inline examples/qwen3-8b-single-node/run.sh:64]: Previously, vLLM hardcodes the config in its yaml ([configs/vllm_qwen3_8b.yaml](https://github.com/lightseekorg/TorchSpec/tree/main/configs/vllm_qwen3_8b.yaml))

File: README.md [hunk 2]
```diff
@@ -112,7 +113,7 @@ pip install -e ".[fa]"
 **vLLM**
 
 ```bash
-./examples/qwen3-8b-single-node/run.sh --config configs/vllm_qwen3_8b.yaml
```
Comment (chungen04) [inline README.md:123]: There is no `--config` field in `run.sh` . See [run.sh](https://github.com/lightseekorg/TorchSpec/blob/main/examples/qwen3-8b-single-node/run.sh) docstring.

File: torchspec/config/inference_config.py [hunk 3]
```diff
@@ -105,6 +105,35 @@ class VllmConfig:
...
+@dataclass
+class TrtllmConfig:
+    """Essential TensorRT-LLM engine configuration.
+
+    Wraps TRT-LLM's PyTorch-backend ``LLM`` in ``SaveHiddenStates`` mode; the
+    TorchSpec patch redirects captured hidden states to Mooncake. Only fields
+    TorchSpec explicitly uses are listed; any other ``LLM`` kwarg can be passed
+    via ``extra_args``.
+
+    Single-node tensor parallelism only (nnodes must be 1); multi-node TP is
```
Comment (chungen04) [inline torchspec/config/inference_config.py:117]: This will be checked in [factory.py](https://github.com/lightseekorg/TorchSpec/pull/121/changes#diff-4788259dae809220eddf0b1a947daeb62d7e1de42c34ea03ec4ccde8adb66332R454).

File: patches/trtllm/v1.3.0rc18/trtllm.patch [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
++            self, scheduled_requests: ScheduledRequests,
++            spec_metadata: "SaveHiddenStatesSpecMetadata") -> None:
++        """Walk context_requests in native order -- how the forward packed the
++        buffer. Sorting by py_batch_idx (= py_seq_slot) would misalign slices.
++        """
++        if local_mpi_rank() != 0:
++            return
++        store = self._ensure_mooncake_store()
++        if store is None:
++            return
```
[the remark this answers is not in the document]
Comment (chungen04) [reply patches/trtllm/v1.3.0rc18/trtllm.patch:116]: Mooncake is required for this framework.

**Comment by chungen04** [comment]
@yubofredwang In this PR the TensorRT-LLM engine was integrated as a backend for inference. Please review and I look forward to suggestions, split PR plans, etc. Also, I did not include tests currently. Please let me know if you suggest writing test in this PR or split one. Thank you!

**Comment by yubofredwang** [comment]
> @yubofredwang In this PR the TensorRT-LLM engine was integrated as a backend for inference. Please review and I look forward to suggestions, split PR plans, etc. Also, I did not include tests currently. Please let me know if you suggest writing test in this PR or split one. Thank you!

thanks for the great work! I will run it end to end to verify

**Review by yubofredwang — APPROVED** [review]
